File: tools/architecture_coverage_validator_v1/schemas.py
from __future__ import annotations

import re
from typing import Any

from .common import (
    ROOT,
    exact_type_registry_symbols,
    load,
    normalize_type_name,
    read,
    require,
    rows,
    top_level_yaml_labels,
    validate_owner_pair,
)

_COMPLETIONS = {"RecipeIdV1", "RecipeBodyV1", "ComparisonAxis", "ProtocolRange", "PackageOpaque"}


def validate_schemas(errors: list[str], topology: dict[str, Any]) -> dict[str, Any]:
    manifest = topology["manifest"]
    modules = topology["modules"]
    packages = topology["packages"]
    schema_doc = load(manifest["schema_registry"])
    recipe_doc = load(manifest["recipe_registry"])
    reason_doc = load(manifest["reason_registry"])

    packet_rows = schema_doc.get("packet")
    if not isinstance(packet_rows, list):
        errors.append("schema packet registry missing")
        packet_rows = []

    schema_names: set[str] = set()
    schema_total = 0
    primitive_families = 0
    primitive_registered: set[str] = set()

    for packet in packet_rows:
        if not isinstance(packet, dict) or not isinstance(packet.get("path"), str):
            errors.append("invalid schema packet entry")
            continue
        path = packet["path"]
        require(errors, (ROOT / path).is_file(), f"missing schema packet {path}")
        if not (ROOT / path).is_file():
            continue
        document = load(path)
        groups = document.get("group")
        if not isinstance(groups, list):
            errors.append(f"{path}: group array missing")
            continue
        packet_count = 0
        for group in groups:
            if not isinstance(group, dict):
                errors.append(f"{path}: invalid group")
                continue
            group_id = group.get("id")
            names = group.get("schemas")
            require(errors, isinstance(names, list) and len(names) > 0, f"{path}:{group_id}: schemas missing")
            if not isinstance(names, list):
                continue
            for name in names:
                require(errors, isinstance(name, str) and name != "", f"{path}:{group_id}: invalid schema name")
                if not isinstance(name, str):
                    continue
                require(errors, name not in schema_names, f"duplicate schema/type name {name}")
                schema_names.add(name)
                packet_count += 1
                if packet.get("id") == "primitives":
                    primitive_registered.add(name)

            source_files = group.get("source_files")
            require(errors, isinstance(source_files, list) and len(source_files) > 0, f"{path}:{group_id}: source files missing")
            combined = ""
            for source in source_files if isinstance(source_files, list) else []:
                require(errors, isinstance(source, str) and (ROOT / source).is_file(), f"{path}:{group_id}: missing schema source {source}")
                if isinstance(source, str) and (ROOT / source).is_file():
                    combined += "\n" + read(source)
            for name in names:
                if isinstance(name, str):
                    require(errors, normalize_type_name(name) in combined, f"{path}:{group_id}: {name} absent from declared sources")

            validate_owner_pair(errors, group.get("shape_owner_package"), group.get("shape_owner_module"), modules, f"schema {group_id}")
            validate_owner_pair(errors, group.get("meaning_owner_package"), group.get("meaning_owner_module"), modules, f"schema {group_id}")
            validate_owner_pair(errors, group.get("state_owner_package"), group.get("state_owner_module"), modules, f"schema {group_id}")
            secondary = group.get("secondary_state_owner_packages", [])
            require(errors, isinstance(secondary, list), f"schema {group_id}: secondary owners must be an array")
            for package in secondary if isinstance(secondary, list) else []:
                require(errors, package in packages, f"schema {group_id}: unknown secondary owner {package}")

        require(errors, document.get("schema_count") == packet_count, f"{path}: schema count mismatch")
        require(errors, packet.get("schema_count") == packet_count, f"{path}: summary schema count mismatch")
        schema_total += packet_count
        families = document.get("primitive_family", [])
        require(errors, isinstance(families, list), f"{path}: primitive_family must be an array")
        primitive_families += len(families) if isinstance(families, list) else 0

    require(errors, schema_total == schema_doc.get("schema_or_registry_count") == 217, "schema/type total must be 217")
    require(errors, schema_doc.get("type_registry_named_symbol_count") == 115, "TYPE_REGISTRY symbol count must be 115")
    require(errors, schema_doc.get("completion_symbol_count") == 5, "type completion count must be 5")
    require(errors, primitive_families == schema_doc.get("canonical_primitive_family_count") == 12, "canonical primitive family count must be 12")

    type_registry_symbols = exact_type_registry_symbols(read("docs/contracts/p00/TYPE_REGISTRY.md"))
    require(errors, len(type_registry_symbols) == 115, f"TYPE_REGISTRY derived symbol count is {len(type_registry_symbols)}, expected 115")
    require(errors, type_registry_symbols | _COMPLETIONS == primitive_registered, f"primitive registry mismatch: {sorted((type_registry_symbols | _COMPLETIONS) ^ primitive_registered)}")

    completion_text = read("docs/contracts/p00/TYPE_COMPLETIONS.md")
    for symbol in sorted(_COMPLETIONS):
        require(errors, symbol in completion_text, f"TYPE_COMPLETIONS missing {symbol}")
    p00_manifest = load("docs/contracts/p00/manifest.toml")
    required_files = p00_manifest.get("required_files")
    require(errors, isinstance(required_files, list) and len(required_files) == 13, "P00 required_files must contain 13 entries")
    require(errors, isinstance(required_files, list) and "TYPE_COMPLETIONS.md" in required_files, "P00 manifest does not include TYPE_COMPLETIONS.md")

    source_schema_files = [
        "docs/contracts/p00/SUPPORT_SCHEMAS.md",
        "docs/contracts/p00/SOURCE_GRAPH.md",
        "docs/contracts/p00/RECIPES.md",
        "docs/contracts/p00/QUERY_AND_RESULTS.md",
        "docs/contracts/p00/RECIPE_RESULTS.md",
        "docs/contracts/p00/PROTOCOL_AND_LIFECYCLE.md",
        "docs/contracts/p00/TYPE_COMPLETIONS.md",
    ]
    source_labels: set[str] = set()
    for path in source_schema_files:
        source_labels.update(top_level_yaml_labels(read(path)))

    recipes_text = read("docs/contracts/p00/RECIPES.md")
    recipe_ids = set(
        re.findall(
            r"^(?:-\s*)?([a-z][a-z0-9_]+@1)(?::)?\s*$",
            recipes_text,
            flags=re.MULTILINE,
        )
    )
    require(errors, recipe_ids == {
        "locate@1",
        "find_text@1",
        "inspect_entity@1",
        "compare_implementations@1",
        "explore_entity@1",
        "corpus_profile@1",
        "corpus_delta@1",
        "provenance@1",
        "compile_exact_scan@1",
        "execute_exact_scan@1",
        "expand_handle@1",
    }, "RECIPES.md exact recipe set mismatch")

    unregistered_labels = {
        label for label in source_labels
        if label not in schema_names and label not in recipe_ids
    }
    require(errors, not unregistered_labels, f"unregistered top-level P00 schema labels: {sorted(unregistered_labels)}")

    recipe_rows = rows(recipe_doc, "recipe", "id")
    require(errors, len(recipe_rows) == 11, "recipe registry must contain 11 recipes")
    require(errors, set(recipe_rows) == recipe_ids, f"recipe registry mismatch: {sorted(set(recipe_rows) ^ recipe_ids)}")
    for recipe_id, row in recipe_rows.items():
        require(errors, row.get("request_schema") in schema_names, f"{recipe_id}: unknown request schema")
        require(errors, row.get("result_schema") in schema_names, f"{recipe_id}: unknown result schema")
        owners = row.get("primary_execution_packages")
        require(errors, isinstance(owners, list) and len(owners) > 0, f"{recipe_id}: execution owners missing")
        for package in owners if isinstance(owners, list) else []:
            require(errors, package in packages, f"{recipe_id}: unknown execution package {package}")

    reason_text = read("docs/contracts/p00/REASON_CODES.md")
    source_reason_codes = set(re.findall(r"^([A-Z][A-Z0-9_]+)$", reason_text, flags=re.MULTILINE))
    registry_reason_codes: set[str] = set()
    expected_sizes = {
        "search_reason_codes": 31,
        "protocol_error_codes": 10,
        "contract_error_codes": 10,
    }
    for key, size in expected_sizes.items():
        values = reason_doc.get(key, {}).get("values")
        require(errors, isinstance(values, list), f"reason registry {key} missing")
        require(errors, isinstance(values, list) and len(values) == size, f"reason registry {key} count mismatch")
        registry_reason_codes.update(values if isinstance(values, list) else [])
    require(errors, source_reason_codes == registry_reason_codes, f"reason code registry mismatch: {sorted(source_reason_codes ^ registry_reason_codes)}")
    require(errors, len(registry_reason_codes) == 51, "reason code total must be 51")

    return {
        "schema_total": schema_total,
        "type_registry_symbols": len(type_registry_symbols),
        "completion_symbols": len(_COMPLETIONS),
        "primitive_families": primitive_families,
        "recipe_count": len(recipe_rows),
        "reason_count": len(registry_reason_codes),
    }

File: xtask/src/architecture_coverage/schemas.rs
//! P00 schema/type, recipe and reason-code closure.

use std::collections::BTreeSet;
use std::path::Path;

use toml::Value;

use super::{
    load::{
        Inputs, integer, load_toml, read_text, require, string,
        string_list, validate_module_ref, validate_owner_pair,
    },
    markdown::{
        exact_type_registry_symbols, normalize_type_name, recipe_ids,
        reason_codes, top_level_yaml_labels,
    },
    topology::{TopologySummary, rows_or_empty},
};

const COMPLETIONS: [&str; 5] = [
    "RecipeIdV1",
    "RecipeBodyV1",
    "ComparisonAxis",
    "ProtocolRange",
    "PackageOpaque",
];

const EXPECTED_RECIPES: [&str; 11] = [
    "locate@1",
    "find_text@1",
    "inspect_entity@1",
    "compare_implementations@1",
    "explore_entity@1",
    "corpus_profile@1",
    "corpus_delta@1",
    "provenance@1",
    "compile_exact_scan@1",
    "execute_exact_scan@1",
    "expand_handle@1",
];

pub(super) struct SchemaSummary {
    pub(super) schema_total: usize,
    pub(super) type_registry_symbols: usize,
    pub(super) completion_symbols: usize,
    pub(super) primitive_families: usize,
    pub(super) recipe_count: usize,
    pub(super) reason_count: usize,
}

pub(super) fn validate(
    root: &Path,
    inputs: &Inputs,
    topology: &TopologySummary,
    errors: &mut Vec<String>,
) -> SchemaSummary {
    let mut schema_names = BTreeSet::new();
    let mut primitive_registered = BTreeSet::new();
    let mut schema_total = 0_usize;
    let mut primitive_families = 0_usize;

    let Some(packets) = inputs.schema_doc.get("packet").and_then(Value::as_array) else {
        errors.push("schema packet registry missing".to_owned());
        return SchemaSummary {
            schema_total: 0,
            type_registry_symbols: 0,
            completion_symbols: COMPLETIONS.len(),
            primitive_families: 0,
            recipe_count: 0,
            reason_count: 0,
        };
    };

    for packet in packets {
        let Some(path) = packet.get("path").and_then(Value::as_str) else {
            errors.push("invalid schema packet entry".to_owned());
            continue;
        };
        require(
            errors,
            root.join(path).is_file(),
            format!("missing schema packet {path}"),
        );
        let document = match load_toml(root, path) {
            Ok(document) => document,
            Err(error) => {
                errors.push(error);
                continue;
            }
        };
        let Some(groups) = document.get("group").and_then(Value::as_array) else {
            errors.push(format!("{path}: group array missing"));
            continue;
        };
        let mut packet_count = 0_usize;
        for group in groups {
            let Some(group_table) = group.as_table() else {
                errors.push(format!("{path}: invalid group"));
                continue;
            };
            let group_id = group_table
                .get("id")
                .and_then(Value::as_str)
                .unwrap_or("<missing>");
            let names = group_table
                .get("schemas")
                .and_then(Value::as_array)
                .and_then(|items| {
                    items
                        .iter()
                        .map(|item| item.as_str().map(str::to_owned))
                        .collect::<Option<Vec<_>>>()
                });
            require(
                errors,
                names.as_ref().is_some_and(|items| !items.is_empty()),
                format!("{path}:{group_id}: schemas missing"),
            );
            let names = names.unwrap_or_default();
            for name in &names {
                require(
                    errors,
                    !name.is_empty(),
                    format!("{path}:{group_id}: invalid schema name"),
                );
                require(
                    errors,
                    schema_names.insert(name.clone()),
                    format!("duplicate schema/type name {name}"),
                );
                if packet.get("id").and_then(Value::as_str) == Some("primitives") {
                    primitive_registered.insert(name.clone());
                }
                packet_count = packet_count.saturating_add(1);
            }

            let source_files = group_table
                .get("source_files")
                .and_then(Value::as_array)
                .and_then(|items| {
                    items
                        .iter()
                        .map(|item| item.as_str().map(str::to_owned))
                        .collect::<Option<Vec<_>>>()
                });
            require(
                errors,
                source_files.as_ref().is_some_and(|items| !items.is_empty()),
                format!("{path}:{group_id}: source files missing"),
            );
            let mut combined = String::new();
            for source in source_files.unwrap_or_default() {
                require(
                    errors,
                    root.join(&source).is_file(),
                    format!("{path}:{group_id}: missing schema source {source}"),
                );
                if let Ok(text) = read_text(root, &source) {
                    combined.push('\n');
                    combined.push_str(&text);
                }
            }
            for name in &names {
                require(
                    errors,
                    combined.contains(normalize_type_name(name)),
                    format!("{path}:{group_id}: {name} absent from declared sources"),
                );
            }

            for (package_key, module_key) in [
                ("shape_owner_package", "shape_owner_module"),
                ("meaning_owner_package", "meaning_owner_module"),
                ("state_owner_package", "state_owner_module"),
            ] {
                validate_owner_pair(
                    errors,
                    group_table.get(package_key).and_then(Value::as_str),
                    group_table.get(module_key).and_then(Value::as_str),
                    &topology.modules,
                    &format!("schema {group_id}"),
                    true,
                );
            }
            for package in group_table
                .get("secondary_state_owner_packages")
                .and_then(Value::as_array)
                .into_iter()
                .flatten()
                .filter_map(Value::as_str)
            {
                require(
                    errors,
                    topology.packages.contains(package),
                    format!("schema {group_id}: unknown secondary owner {package}"),
                );
            }
        }
        require(
            errors,
            integer(&document, "schema_count") == i64::try_from(packet_count).ok(),
            format!("{path}: schema count mismatch"),
        );
        require(
            errors,
            packet.get("schema_count").and_then(Value::as_integer)
                == i64::try_from(packet_count).ok(),
            format!("{path}: summary schema count mismatch"),
        );
        schema_total = schema_total.saturating_add(packet_count);
        primitive_families = primitive_families.saturating_add(
            document
                .get("primitive_family")
                .and_then(Value::as_array)
                .map_or(0, Vec::len),
        );
    }

    require(
        errors,
        schema_total == 217
            && integer(&inputs.schema_doc, "schema_or_registry_count") == Some(217),
        "schema/type total must be 217",
    );
    require(
        errors,
        integer(&inputs.schema_doc, "type_registry_named_symbol_count") == Some(115),
        "TYPE_REGISTRY symbol count must be 115",
    );
    require(
        errors,
        integer(&inputs.schema_doc, "completion_symbol_count") == Some(5),
        "type completion count must be 5",
    );
    require(
        errors,
        primitive_families == 12
            && integer(&inputs.schema_doc, "canonical_primitive_family_count") == Some(12),
        "canonical primitive family count must be 12",
    );

    let type_registry_symbols = match read_text(root, "docs/contracts/p00/TYPE_REGISTRY.md") {
        Ok(text) => match exact_type_registry_symbols(&text) {
            Ok(symbols) => symbols,
            Err(error) => {
                errors.push(error);
                BTreeSet::new()
            }
        },
        Err(error) => {
            errors.push(error);
            BTreeSet::new()
        }
    };
    require(
        errors,
        type_registry_symbols.len() == 115,
        format!(
            "TYPE_REGISTRY derived symbol count is {}, expected 115",
            type_registry_symbols.len()
        ),
    );
    let completion_symbols: BTreeSet<String> =
        COMPLETIONS.iter().map(|value| (*value).to_owned()).collect();
    let expected_primitives: BTreeSet<String> = type_registry_symbols
        .union(&completion_symbols)
        .cloned()
        .collect();
    require(
        errors,
        expected_primitives == primitive_registered,
        format!(
            "primitive registry mismatch: {:?}",
            expected_primitives
                .symmetric_difference(&primitive_registered)
                .cloned()
                .collect::<Vec<_>>()
        ),
    );

    validate_completion_contract(root, &inputs.p00_manifest, errors);
    let recipe_count = validate_recipes(
        root,
        inputs,
        topology,
        &schema_names,
        errors,
    );
    let reason_count = validate_reasons(root, &inputs.reason_doc, errors);

    SchemaSummary {
        schema_total,
        type_registry_symbols: type_registry_symbols.len(),
        completion_symbols: completion_symbols.len(),
        primitive_families,
        recipe_count,
        reason_count,
    }
}

fn validate_completion_contract(
    root: &Path,
    p00_manifest: &Value,
    errors: &mut Vec<String>,
) {
    let completion_text = read_text(root, "docs/contracts/p00/TYPE_COMPLETIONS.md")
        .unwrap_or_else(|error| {
            errors.push(error);
            String::new()
        });
    for symbol in COMPLETIONS {
        require(
            errors,
            completion_text.contains(symbol),
            format!("TYPE_COMPLETIONS missing {symbol}"),
        );
    }
    let required_files = string_list(p00_manifest, "required_files");
    require(
        errors,
        required_files.as_ref().is_some_and(|files| files.len() == 13),
        "P00 required_files must contain 13 entries",
    );
    require(
        errors,
        required_files
            .as_ref()
            .is_some_and(|files| files.iter().any(|file| file == "TYPE_COMPLETIONS.md")),
        "P00 manifest does not include TYPE_COMPLETIONS.md",
    );
}

fn validate_recipes(
    root: &Path,
    inputs: &Inputs,
    topology: &TopologySummary,
    schema_names: &BTreeSet<String>,
    errors: &mut Vec<String>,
) -> usize {
    let source_files = [
        "docs/contracts/p00/SUPPORT_SCHEMAS.md",
        "docs/contracts/p00/SOURCE_GRAPH.md",
        "docs/contracts/p00/RECIPES.md",
        "docs/contracts/p00/QUERY_AND_RESULTS.md",
        "docs/contracts/p00/RECIPE_RESULTS.md",
        "docs/contracts/p00/PROTOCOL_AND_LIFECYCLE.md",
        "docs/contracts/p00/TYPE_COMPLETIONS.md",
    ];
    let mut source_labels = BTreeSet::new();
    for relative in source_files {
        match read_text(root, relative) {
            Ok(text) => source_labels.extend(top_level_yaml_labels(&text)),
            Err(error) => errors.push(error),
        }
    }
    let recipes_text = read_text(root, "docs/contracts/p00/RECIPES.md")
        .unwrap_or_else(|error| {
            errors.push(error);
            String::new()
        });
    let source_recipes = recipe_ids(&recipes_text);
    let expected: BTreeSet<String> =
        EXPECTED_RECIPES.iter().map(|value| (*value).to_owned()).collect();
    require(
        errors,
        source_recipes == expected,
        "RECIPES.md exact recipe set mismatch",
    );
    let unregistered: BTreeSet<String> = source_labels
        .into_iter()
        .filter(|label| !schema_names.contains(label) && !source_recipes.contains(label))
        .collect();
    require(
        errors,
        unregistered.is_empty(),
        format!("unregistered top-level P00 schema labels: {unregistered:?}"),
    );

    let rows = rows_or_empty(&inputs.recipe_doc, "recipe", "id", errors);
    require(
        errors,
        rows.len() == 11,
        "recipe registry must contain 11 recipes",
    );
    require(
        errors,
        rows.keys().cloned().collect::<BTreeSet<_>>() == source_recipes,
        "recipe registry mismatch",
    );
    for (recipe, row) in &rows {
        require(
            errors,
            string(row, "request_schema").is_some_and(|name| schema_names.contains(name)),
            format!("{recipe}: unknown request schema"),
        );
        require(
            errors,
            string(row, "result_schema").is_some_and(|name| schema_names.contains(name)),
            format!("{recipe}: unknown result schema"),
        );
        let owners = string_list(row, "primary_execution_packages");
        let refs = string_list(row, "execution_modules");
        require(
            errors,
            owners.as_ref().is_some_and(|items| !items.is_empty()),
            format!("{recipe}: execution owners missing"),
        );
        require(
            errors,
            refs.as_ref().zip(owners.as_ref()).is_some_and(|(refs, owners)| refs.len() == owners.len()),
            format!("{recipe}: one execution module per owner required"),
        );
        let mut referenced_packages = BTreeSet::new();
        for reference in refs.unwrap_or_default() {
            validate_module_ref(errors, &reference, &topology.modules, recipe);
            if let Some((package, _)) = reference.split_once(':') {
                referenced_packages.insert(package.to_owned());
            }
        }
        let owner_set: BTreeSet<String> = owners.unwrap_or_default().into_iter().collect();
        for package in &owner_set {
            require(
                errors,
                topology.packages.contains(package),
                format!("{recipe}: unknown execution package {package}"),
            );
        }
        require(
            errors,
            referenced_packages == owner_set,
            format!("{recipe}: execution module/package mismatch"),
        );
    }
    rows.len()
}

fn validate_reasons(
    root: &Path,
    reason_doc: &Value,
    errors: &mut Vec<String>,
) -> usize {
    let source = read_text(root, "docs/contracts/p00/REASON_CODES.md")
        .map(|text| reason_codes(&text))
        .unwrap_or_else(|error| {
            errors.push(error);
            BTreeSet::new()
        });
    let mut registered = BTreeSet::new();
    for (key, expected_count) in [
        ("search_reason_codes", 31_usize),
        ("protocol_error_codes", 10),
        ("contract_error_codes", 10),
    ] {
        let values = reason_doc
            .get(key)
            .and_then(Value::as_table)
            .and_then(|table| table.get("values"))
            .and_then(Value::as_array)
            .and_then(|items| {
                items
                    .iter()
                    .map(|item| item.as_str().map(str::to_owned))
                    .collect::<Option<Vec<_>>>()
            });
        require(
            errors,
            values.is_some(),
            format!("reason registry {key} missing"),
        );
        require(
            errors,
            values.as_ref().is_some_and(|items| items.len() == expected_count),
            format!("reason registry {key} count mismatch"),
        );
        registered.extend(values.unwrap_or_default());
    }
    require(
        errors,
        source == registered,
        format!(
            "reason code registry mismatch: {:?}",
            source
                .symmetric_difference(&registered)
                .cloned()
                .collect::<Vec<_>>()
        ),
    );
    require(
        errors,
        registered.len() == 51,
        "reason code total must be 51",
    );
    registered.len()
}
